### ISSUE TITLE
LGBMRegressor always enable L2 metric, which makes 'first_metric_only' not working

### ISSUE THREAD

**Issue by HarryAtDelphia** [microsoft/LightGBM#4549]
## Description

I would like to train a LGBMRegressor and evaluate on correlation, however I noticed that LGBMRegressor always has the L2 as the first evaluation metric, and evaluated on L2 only.

## Reproducible example

```
def corr_eval(y, data):
    corr = np.corrcoef(y, data)
    return 'corr', corr[0, 1], True

report_df = {}
model_config = {'alpha': 0.9,
                'colsample_bytree': 0.5,
                'learning_rate': 0.01,
                'n_estimators': 2000,
                'max_bin': 255,
                'max_depth': 25,
                'min_child_samples': 10,
                'min_data_in_leaf': 10,
                'min_split_gain': 0.003,
                'num_leaves': 25,
                'objective': 'regression_l2',
                'random_state': 0,
                'reg_alpha': 0,
                'reg_lambda': 0,
                'subsample': 0.7,
                'subsample_freq': 3,
                'random_state': 123,}
model = LGBMRegressor(metrics=None, first_metric_only=True, **model_config)
model.fit(x_tr, y_tr, categorical_feature=categorical_features, feature_name=features,
          eval_set=[(valid_x, valid_y)], eval_metric=[corr_eval], eval_names=['valid'],early_stopping_rounds=600,
         )

```
## Environment info

LightGBM version or commit hash: lightgbm        3.1.1        py38h23ab428_0 

Command(s) you used to install LightGBM

```shell

```

## Additional Comments

![image](https://user-images.githubusercontent.com/79157811/130523178-43ce0ea5-0111-4f4b-a1b2-12772c86fa65.png)
![image](https://user-images.githubusercontent.com/79157811/130523195-3464e6af-8785-4978-91e6-07571d6d337e.png)
Clearly, if evaluate on correlation function, round 570-600 are higher than round 2, although I set the corr_eval as the evaluation metric, it plots for L2 and evaluate on L2 only.

**Comment by jameslamb** [comment]
Hi @HarryAtDelphia , thanks for using LightGBM!

> round 570-600 are higher than round 2, although I set the corr_eval as the evaluation metric

I think you may have misunderstood how the parameter `early_stopping_rounds` ([documented here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#early_stopping_round)) works.

Setting `early_stopping_rounds=600` means "stop training after 600 consecutive iterations where performance on a validation set failed to improve". So comparing iteration 2 to iteration 600 is not sufficient to show that early stopping is / is not working.

Also note that as written, it seems like the `corr_eval` metric _is improving_ as training proceeds. Since you returned `True` for the `higher_better` component of `corr_eval()` (the third item in its return tuple), you are telling LightGBM "less negative values are better". So LightGBM will think that values like -0.05 are better than values like -0.19 because they are "higher". If you want to measure the strength of correlation regardless of its direction and perform early stopping based on that, you probably need to take the absolute value of correlation (using something like `numpy.abs()`).

-------

If I've misunderstood and/or you'd like additional help, please update the sample code to be a fully-reproducible example (something that maintainers can copy and run with no additional modification required).

That means:

* including all relevant `import` statements
* including the data-loading code that defines `x_tr`, `y_tr`, and other data structures (the code in `scklearn.datasets` from `scikit-learn` is useful for this purpose)

**Comment by StrikerRUS** [comment]
@HarryAtDelphia If you'd like to ignore default metric associated with objective function, you should pass `metrics="None"` as a string.

```
...
model = lgb.LGBMRegressor(metrics='None', first_metric_only=True, **model_config)
...
```

**Comment by HarryAtDelphia** [comment]
> Hi @HarryAtDelphia , thanks for using LightGBM!
> 
> > round 570-600 are higher than round 2, although I set the corr_eval as the evaluation metric
> 
> I think you may have misunderstood how the parameter `early_stopping_rounds` ([documented here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#early_stopping_round)) works.
> 
> Setting `early_stopping_rounds=600` means "stop training after 600 consecutive iterations where performance on a validation set failed to improve". So comparing iteration 2 to iteration 600 is not sufficient to show that early stopping is / is not working.
> 
> Also note that as written, it seems like the `corr_eval` metric _is improving_ as training proceeds. Since you returned `True` for the `higher_better` component of `corr_eval()` (the third item in its return tuple), you are telling LightGBM "less negative values are better". So LightGBM will think that values like -0.05 are better than values like -0.19 because they are "higher". If you want to measure the strength of correlation regardless of its direction and perform early stopping based on that, you probably need to take the absolute value of correlation (using something like `numpy.abs()`).
> 
> If I've misunderstood and/or you'd like additional help, please update the sample code to be a fully-reproducible example (something that maintainers can copy and run with no additional modification required).
> 
> That means:
> 
> * including all relevant `import` statements
> * including the data-loading code that defines `x_tr`, `y_tr`, and other data structures (the code in `scklearn.datasets` from `scikit-learn` is useful for this purpose)

Thank you for help. I am sorry that I didn't make it clear. Exactly as you said it seemed 'corr_eval' metric was improving as training proceeded, so if the early stopping is evaluated based on this metric, the training shouldn't stop at round 600 but keep going on. The fact the training stopped at round 602 and reported round 2 as the best is because it used 'L2' as evaluation metric. So I would like the model ignore the default 'L2' metric and evaluate early stopping based on 'corr_eval' instead.

**Comment by HarryAtDelphia** [comment]
> Hi @HarryAtDelphia , thanks for using LightGBM!
> 
> > round 570-600 are higher than round 2, although I set the corr_eval as the evaluation metric
> 
> I think you may have misunderstood how the parameter `early_stopping_rounds` ([documented here](https://lightgbm.readthedocs.io/en/latest/Parameters.html#early_stopping_round)) works.
> 
> Setting `early_stopping_rounds=600` means "stop training after 600 consecutive iterations where performance on a validation set failed to improve". So comparing iteration 2 to iteration 600 is not sufficient to show that early stopping is / is not working.
> 
> Also note that as written, it seems like the `corr_eval` metric _is improving_ as training proceeds. Since you returned `True` for the `higher_better` component of `corr_eval()` (the third item in its return tuple), you are telling LightGBM "less negative values are better". So LightGBM will think that values like -0.05 are better than values like -0.19 because they are "higher". If you want to measure the strength of correlation regardless of its direction and perform early stopping based on that, you probably need to take the absolute value of correlation (using something like `numpy.abs()`).
> 
> If I've misunderstood and/or you'd like additional help, please update the sample code to be a fully-reproducible example (something that maintainers can copy and run with no additional modification required).
> 
> That means:
> 
> * including all relevant `import` statements
> * including the data-loading code that defines `x_tr`, `y_tr`, and other data structures (the code in `scklearn.datasets` from `scikit-learn` is useful for this purpose)

**Comment by HarryAtDelphia** [comment]
> @HarryAtDelphia If you'd like to ignore default metric associated with objective function, you should pass `metrics="None"` as a string.
> 
> ```
> ...
> model = lgb.LGBMRegressor(metrics='None', first_metric_only=True, **model_config)
> ...
> ```

The string works! Thank you for help.
